Fix WebRequest failure to handle missing ContentType in response header

nmbradford opened the pull request:
If a response to invoke-webrequest does not set the content type in the
response header, an object not set exception is thrown in
WebResponseHelper.CoreClr.cs.  Updated to set contenttype = null.

Commit:
Fix WebRequest to handle missing ContentType in response header

If a response to invoke-webrequest does not set the content type in the
response header, an object not set exception is thrown in
WebResponseHelper.CoreClr.cs.  Updated to set contenttype = null.

msftclas commented:
Hi __@nmbradford__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -351,7 +351,18 @@ protected override void ProcessRecord()
+                    string contentType = null;
+                    try 
+                    { 
+                        ContentHelper.GetContentType(response);
```
Maybe `contentType = ContentHelper.GetContentType(response); `

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:
```diff
@@ -38,7 +38,17 @@ internal static string GetStatusDescription(HttpResponseMessage response)
+            }
+            catch 
+            {
+
```
Please remove unneeded empty line.

iSazonov reviewed (commented):
@nmbradford  You would need to add tests.

Commit:
Resolve review issues

Removed blank line and added missing assignment for contentType

Commit:
Added test for missing Content-Type header

nmbradford commented:
I've added a single test for the exception that is thrown by iwr if Content-Type is missing from the response header and addressed the issues in review.  

Also - I have signed the CLA,. but the cla-required label is still attached to this issue?

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -391,6 +391,14 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+        $result.Error | Should BeNullOrEmpty
+
+    }
 
```
Please remove unneeded empty lines L400 and L402.

iSazonov commented:
@nmbradford Thanks! LGTM (Only correct the test). 
Wait for maintainers to review and remove cla.

Commit:
Removed unneeded blank lines

nmbradford commented:
done - thanks @iSazonov

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -351,7 +351,17 @@ protected override void ProcessRecord()
+                    { 
+                        contentType = ContentHelper.GetContentType(response);
+                    }
+                    catch {
```
Catch should be for a specific exception type

nmbradford replied:
add explicit catch for NullReferenceException

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:
```diff
@@ -38,7 +38,16 @@ internal static string GetStatusDescription(HttpResponseMessage response)
+            try { 
+                contentType = response.Content.Headers.ContentType.MediaType;
+            }
+            catch 
```
Same: catch should be for a specific exception type

nmbradford replied:
added explicit catch for NullReferenceException

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -351,7 +351,17 @@ protected override void ProcessRecord()
 
-                    string contentType = ContentHelper.GetContentType(response);
+                    //added try/catch due to possible lack of content type in the response header.
+                    //seems the only reason we are getting it here is for a log message?
```
Current comments are pretty vague and suggest that author of the code is not completely sure about what the code does. It's totally ok for a codereview discussion, maybe even test code, but it should be avoided in the product code.

Think about the comments from the next-person-who-is-going-to-edit-it perspective.
The purpose of the comments is to **clarify** some aspects of the code, not to admit that code may be written incorrectly.

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -351,7 +351,17 @@ protected override void ProcessRecord()
+                        contentType = ContentHelper.GetContentType(response);
+                    }
+                    catch {
+                        contentType = null;
```
redundant, `contentType` is already assigned to null, on declaration

nmbradford replied:
Removed

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:
```diff
@@ -38,7 +38,16 @@ internal static string GetStatusDescription(HttpResponseMessage response)
+            }
+            catch 
+            {
+                contentType = null;
```
Same, redundent, because already assigned to null

nmbradford replied:
Removed as well

Commit:
Clarify comments and removed redundant assignment to null

nmbradford commented:
Thanks @vors - have made suggested changes...

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -351,7 +351,18 @@ protected override void ProcessRecord()
+                    { 
+                        contentType = ContentHelper.GetContentType(response);
+                    }
+                    catch (NullReferenceException) 
```
NullReferenceException happens only when you are trying to de-reference null object. Exceptions are for exception situations and it's pretty easy to avoid NullReferenceException: we just need to check for `null` in the appropriate place. In this case, in `src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs` GetContentType.
There is a long chain of properties: `response.Content.Headers.ContentType.MediaType`. I don't know from the top of my head, which one is `null`, in this case, but I think you know.
Also that way we really fixing the problem in the implementation and not only on the caller side.

vors replied:
That also means, that comment above should also be moved to the `ContentHelper.CoreClr.cs`

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:
```diff
@@ -38,7 +39,15 @@ internal static string GetStatusDescription(HttpResponseMessage response)
+            try { 
+                contentType = response.Content.Headers.ContentType.MediaType;
+            }
+            catch (NullReferenceException) 
```
same

Commit:
Test for ContentType header before accessing properties

Deal with missing ContentType in response header at point of access rather than
by catching NullRef in caller to address previous commit review concerns.

Realised that similar issue affects invoke-restmethod.  Added test to cause nullref ex
to be thrown when invoke-restmethod's use causes a response with no ContentType in the
response header, and added fix in ContentHelper.GetEncoding.

Commit:
Removed unneeded reference to System introduced in previous commit

Commit:
Removed spaces

nmbradford commented:
Thanks for the feedback @vors.  I've updated the PR to prevent the nullref instead of catching.  I also realised the same issue occurs in Invoke-RestMethod and the fix is almost identical, so I added a test for it and address it as well.  Let me know if this needs to go in a separate PR.

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs:
```diff
@@ -14,13 +14,26 @@ internal static partial class ContentHelper
+            if ( response.Content.Headers.ContentType != null)
+            { 
+                charSet = response.Content.Headers.ContentType.CharSet;
+            }
```
C# 6 has a nice sugar for this, [Null-Conditional Operator](https://msdn.microsoft.com/en-us/magazine/dn802602.aspx)

```csharp
string charSet = response.Content.Headers.ContentType?.CharSet;
```

nmbradford replied:
Ahh - that's nice and concise!  Thanks! Did you want me to update the PR with changes to that effect, or is just FYI (which is definitely appreciated since I'm a PowerShell guy who knows just enough csharp to hang himself ;) )

vors replied:
Yes, can you, please, update the code?

nmbradford replied:
done - back to a three character fix :)  Consider me schooled, thanks @vors

vors reviewed (changes requested):
Thank you, it's great that you approach problem more generic and fixing `Invoke-RestMethod` at the same time!

Commit:
Update check for null to use null condition operator

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/ContentHelper.CoreClr.cs:
```diff
@@ -14,13 +14,18 @@ internal static partial class ContentHelper
         {
-            string charSet = response.Content.Headers.ContentType.CharSet;
+                
+            //ContentType may not exist in response header.
```
In the future it might not pass spelling verification. It would be nice to insert space after `//`
`// ContentType may not exist in response header.  Return null if not.`
And in below comments too.

nmbradford replied:
Added.

Commit:
Added space in comment

Commit:
Fix blank lines

vors commented:
I updated the PR to remove extra blank lines

msftclas commented:
Hi __@nmbradford__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

vors commented:
CLA automation failed to remove `cla-required`, but CLA is signed.

iSazonov commented:
@vors Maybe it is collision with @nmbradford mails?

iSazonov commented:
@nmbradford Thanks for the successful completion ot the fix!
